fqdn/namemanager: Pre-allocate Identities for FQDN Selectors

nathanjsweet opened the pull request:
This is a performance improvement for DNS lookups.
Critical path latency for DNS lookups is reduced
by pre-allocating a matching identity whenever
the policy engine adds a new ToFQDN selector,
because no policy maps need to be updated when 
the first DNS lookup occurs.

nathanjsweet committed:
fqdn/namemanager: Pre-allocate Identities for FQDN Selectors

This is a performance improvement for DNS lookups.
Critical path latency for DNS lookups is reduced
by pre-allocating a matching identity whenever
the policy engine adds a new ToFQDN selector,
because no policy maps need to be updated when
the first DNS lookup occurs.

Co-authored-by: David Bimmler <david.bimmler@isovalent.com>
Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

christarazi commented on pkg/fqdn/namemanager/manager.go:
```diff
@@ -109,6 +111,10 @@ func (n *manager) RegisterFQDNSelector(selector api.FQDNSelector) {
+		// Required to be asynchronous to avoid a deadlock. The selectorcache calls
+		// namemanager while being locked, and allocating an identity with notifyOwner calls
+		// into the selector cache.
+		go n.allocateIdentity(selector)
```
The unboundedness here of launching goroutines can get us into trouble. We have a few options here:

1. Bound via a workerpool of size `runtime.NumCPU()`
2. Bound via a `trigger.Trigger` (single threaded worker that handles identities and is triggered from here)

I'm leaning more towards (2) since it feels like (1) is slightly overkill and (2) is sufficient.

squeed commented on pkg/fqdn/namemanager/manager.go:
```diff
@@ -467,3 +487,70 @@ func (n *manager) mapSelectorsToNamesLocked(fqdnSelector api.FQDNSelector) (name
+	for _, sel := range n.allocSelectors {
+		lbls := identityLabels(sel)
+		for _, ls := range lbls {
+			_, _, err := n.params.IdentityAllocator.AllocateIdentity(ctx, ls, true, identity.InvalidIdentity)
```
You can use `AllocateLocalIdentity()` which doesn't require a context. It just needs to be added to the `IdentityAllocator` interface.

nathanjsweet replied:
`AllocateLocalIdentity` is not part of the `IdentityAllocator` interface. It looks like  `AllocateIdentity` immediately checks the labels. `AllocateIdentities` immediately calls `needsGlobalIdentity` to see if the labels can be locally allocated. I would prefer to use that logic than introduce a new method into the `IdentityAllocator` interface.

nathanjsweet committed:
fqdn/namemanager: Add Tests for Pre Allocation

The TestNameManagerIPCacheUpdates test
now has an identity cache that has its changes saved
locally to compare with the labels that are expected
from newly generated fqdn selectors.

The method also had its Equals checks for expected vs
actual reversed, this is rectified in this commit.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
